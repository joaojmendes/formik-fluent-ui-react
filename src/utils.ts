import { FieldProps, getIn } from 'formik'

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export function getErrorMessage<T = any>({ field, form }: FieldProps<T>) {
  const error = getIn(form.errors, field.name)
  const touched = getIn(form.touched, field.name)

  return touched ? error : undefined
}

export function invokeAll(...callbacks: any[]) {
  return () => {
    for (const callback of callbacks) {
      if (callback && typeof callback === 'function') {
        callback()
      }
    }
  }
}

export function createFakeEvent({ name }: { name: string }) {
  return { target: { name } }
}
