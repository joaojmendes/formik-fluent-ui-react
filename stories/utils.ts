import { action } from '@storybook/addon-actions'
import { FormikHelpers } from 'formik'

export function handleSubmit<T>(
  values: T,
  { setSubmitting }: FormikHelpers<T>
) {
  action('submit:start')(values)
  setTimeout(() => {
    setSubmitting(false)
    action('submit:done')()
  }, 2000)
}
