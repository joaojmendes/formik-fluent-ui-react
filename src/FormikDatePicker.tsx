import { DatePicker, IDatePickerProps } from '@fluentui/react';
import { FieldProps } from 'formik'
import * as React from 'react'
import { createFakeEvent, invokeAll, Omit } from './utils'

export function mapFieldToDatePicker<T = any>({
  form,
  field,
}: FieldProps<T>): Pick<
  IDatePickerProps,
  'value' | 'onSelectDate' | 'onAfterMenuDismiss'
> {
  return {
    value: field.value,
    onAfterMenuDismiss: () => field.onBlur(createFakeEvent(field)),
    onSelectDate: date => form.setFieldValue(field.name, date),
  }
}
export type FormikDatePickerProps<T = any> = Omit<
  IDatePickerProps,
  'value' | 'onSelectDate' | 'onBlur' | 'onChange'
> &
  FieldProps<T>

export function FormikDatePicker<T = any>({
  field,
  form,
  ...props
}: FormikDatePickerProps<T>) {
  const { onAfterMenuDismiss, ...fieldProps } = mapFieldToDatePicker({
    field,
    form,
  })

  return (
    <DatePicker
      {...props}
      onAfterMenuDismiss={invokeAll(
        onAfterMenuDismiss,
        props.onAfterMenuDismiss
      )}
      {...fieldProps}
    />
  )
}
