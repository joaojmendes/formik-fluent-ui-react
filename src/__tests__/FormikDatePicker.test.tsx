import { Field, FieldProps, Form, Formik } from 'formik'
import { DatePicker, setIconOptions } from 'office-ui-fabric-react'
import * as React from 'react'
import renderer from 'react-test-renderer'
import { FormikDatePicker, mapFieldToDatePicker } from '../FormikDatePicker'
import { noop, serialize } from './utils'

// Suppress icon warnings.
setIconOptions({
  disableWarnings: true,
})

const testDate = new Date('2018-12-04T12:00:00Z')

class Values {
  public date: Date = testDate
}

function createFieldProps(value = testDate): FieldProps<Values> {
  return {
    field: {
      value,
      onChange: jest.fn(),
      onBlur: jest.fn(),
      name: 'date',
    },
    form: { setFieldValue: jest.fn(), handleBlur: jest.fn(() => jest.fn()) },
  } as any
}

test('<FormikDatePicker /> renders correctly as a field component', () => {
  const component = renderer.create(
    <Formik initialValues={{ test: testDate }} onSubmit={noop}>
      <Form>
        <Field name="date" label="Date" component={FormikDatePicker} />
      </Form>
    </Formik>
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('<FormikDatePicker /> renders a Fabric <DatePicker />', () => {
  const fieldProps = createFieldProps()

  const formikDatePicker = renderer.create(
    <FormikDatePicker {...fieldProps} label="Date" />
  )
  const fabricDatePicker = renderer.create(
    <DatePicker {...mapFieldToDatePicker(fieldProps)} label="Date" />
  )
  expect(serialize(formikDatePicker)).toBe(serialize(fabricDatePicker))
})

test('mapFieldToDatePicker() maps FieldProps to IDatePickerProps', () => {
  const { field, form } = createFieldProps()
  const props = mapFieldToDatePicker({ form, field })

  expect(props.value).toBe(field.value)

  const selectedDate = new Date(testDate.getTime() + 86400 * 1000) // tomorrow
  props.onSelectDate!(selectedDate)

  expect(form.setFieldValue).toHaveBeenCalledTimes(1)
  expect(form.setFieldValue).toHaveBeenCalledWith(field.name, selectedDate)
  expect(field.onChange).not.toHaveBeenCalled()
})
