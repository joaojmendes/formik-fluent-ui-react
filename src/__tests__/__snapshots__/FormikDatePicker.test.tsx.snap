// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormikDatePicker /> renders correctly as a field component 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="ms-DatePicker root-53"
  >
    <div
      aria-haspopup="true"
    >
      <div
        className="ms-TextField textField-57"
      >
        <div
          className="ms-TextField-wrapper"
        >
          <label
            className="ms-Label root-68"
            htmlFor="DatePicker0-label"
            id="TextFieldLabel4"
          >
            Date
          </label>
          <div
            className="ms-TextField-fieldGroup fieldGroup-58"
          >
            <input
              aria-expanded={false}
              aria-invalid={false}
              aria-labelledby="TextFieldLabel4"
              className="ms-TextField-field field-59"
              id="DatePicker0-label"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onInput={[Function]}
              onKeyDown={[Function]}
              readOnly={true}
              required={false}
              role="combobox"
              type="text"
              value="Tue Dec 04 2018"
            />
            <i
              aria-hidden={true}
              className="ms-DatePicker-event--with-label msDatePickerDisabled icon-69"
              data-icon-name="Calendar"
              onClick={[Function]}
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;
